### PR TITLE
add support for a config.d directory

### PR DESCRIPTION
**Pull request by marcial-lopezferrada-hs**

### CODE REVIEW

File: config/config.go [hunk 1]
```diff
@@ -168,7 +167,36 @@ func ReadConfigFile(configFile string, inputPrefix, outputPrefix string) (*Contr
...
-	return ReadConfig(log, yamlFile, inputPrefix, outputPrefix)
+	config, err := ReadConfig(log, yamlFile, inputPrefix, outputPrefix)
+
+	items, _ := ioutil.ReadDir(util.AbsolutePath(inputPrefix, configDir))
+	for _, item := range items {
+		fileExtension := filepath.Ext(item.Name())
+		if fileExtension == ".yaml" || fileExtension == ".yml" {
+
+			yamlFile, err := ioutil.ReadFile(util.AbsolutePath(inputPrefix, configDir) + "/" + item.Name())
+			current_config, err := ReadConfig(log, yamlFile, inputPrefix, outputPrefix)
```
Comment (tom-hadlaw-hs) [inline config/config.go:178]: currentConfig would be preferred

File: config/config.go [hunk 2]
```diff
@@ -168,7 +167,36 @@ func ReadConfigFile(configFile string, inputPrefix, outputPrefix string) (*Contr
...
+			config.VaultConfig.SSHCertificates = append(config.VaultConfig.SSHCertificates, current_config.VaultConfig.SSHCertificates...)
+			config.VaultConfig.AWS = append(config.VaultConfig.AWS, current_config.VaultConfig.AWS...)
+			for k, v := range current_config.Templates {
+				config.Templates[k] = v
+			}
+			for k, v := range current_config.Composites {
+				config.Composites[k] = v
+			}
+
+			if err != nil {
```
Comment (tom-hadlaw-hs) [inline config/config.go:190]: Why don't we check this right after the call to read config?

File: config/config.go [hunk 3]
```diff
@@ -151,7 +150,7 @@ func ReadConfig(log zerolog.Logger, config []byte, inputPrefix, outputPrefix str
 		Composites:  composites,
 	}, nil
 }
-func ReadConfigFile(configFile string, inputPrefix, outputPrefix string) (*ControlToolConfig, error) {
+func ReadConfigFile(configFile string, configDir string, inputPrefix, outputPrefix string) (*ControlToolConfig, error) {
 	if configFile == "" {
 		return nil, fmt.Errorf("a --config file is required to be specified")
```
Comment (tom-hadlaw-hs) [inline config/config.go:155]: If we specify a configDir, do we need to specify a configFile.

File: config/config.go [hunk 4]
```diff
@@ -168,7 +167,36 @@ func ReadConfigFile(configFile string, inputPrefix, outputPrefix string) (*Contr
...
 	}
 
-	return ReadConfig(log, yamlFile, inputPrefix, outputPrefix)
+	config, err := ReadConfig(log, yamlFile, inputPrefix, outputPrefix)
+
+	items, _ := ioutil.ReadDir(util.AbsolutePath(inputPrefix, configDir))
+	for _, item := range items {
+		fileExtension := filepath.Ext(item.Name())
+		if fileExtension == ".yaml" || fileExtension == ".yml" {
+
```
Comment (tom-hadlaw-hs) [inline config/config.go:176]: Adding some debug logging for files that are read into config might be worthwhile for debugging.

File: config/config.go [hunk 5]
```diff
@@ -168,7 +167,36 @@ func ReadConfigFile(configFile string, inputPrefix, outputPrefix string) (*Contr
 		return nil, fmt.Errorf("trouble reading config file %q: %w", absConfigFile, err)
 	}
 
-	return ReadConfig(log, yamlFile, inputPrefix, outputPrefix)
+	config, err := ReadConfig(log, yamlFile, inputPrefix, outputPrefix)
```
Comment (tom-hadlaw-hs) [inline config/config.go:170]: Error doesn't get checked here

File: config/config.go [hunk 6]
```diff
@@ -168,7 +167,36 @@ func ReadConfigFile(configFile string, inputPrefix, outputPrefix string) (*Contr
...
+
+	items, _ := ioutil.ReadDir(util.AbsolutePath(inputPrefix, configDir))
```
Comment (tom-hadlaw-hs) [inline config/config.go:172]: We should handle the error ReadDir returns.

File: config/config.go [hunk 7]
```diff
@@ -168,7 +167,36 @@ func ReadConfigFile(configFile string, inputPrefix, outputPrefix string) (*Contr
...
+	for _, item := range items {
+		fileExtension := filepath.Ext(item.Name())
+		if fileExtension == ".yaml" || fileExtension == ".yml" {
```
Comment (tom-hadlaw-hs) [inline config/config.go:175]: Probably worth to do

```suggestion
		if (fileExtension == ".yaml" || fileExtension == ".yml") && !item.IsDir() {
```

File: config/config.go (hunk 5 above)
Comment (tom-hadlaw-hs) [reply config/config.go:170]: I wonder if the specified file should have the highest precedence.

File: perform.go [hunk 8]
```diff
@@ -3,12 +3,13 @@ package main
 import (
 	"context"
 	"fmt"
-	"github.com/hootsuite/vault-ctrl-tool/v2/metrics"
 	"os"
 	"os/signal"
 	"syscall"
 	"time"
 
+	"github.com/hootsuite/vault-ctrl-tool/v2/metrics"
```
Comment (tom-hadlaw-hs) [inline perform.go:11]: Can you group up the internal imports

File: config/config.go (hunk 3 above)
Comment (marcial-lopezferrada-hs) [reply config/config.go:155]: I don't think so, but I fear that just having config-dir would break stuff. I'm not sure how we deploy vault-ctrl-tool around, but I can see that we would have instance still using the `vault-config.yml` and not the directory

File: config/config_test.go [hunk 9]
```diff
@@ -63,14 +65,31 @@ func TestInvalidConfigs(t *testing.T) {
...
 		})
 	}
 }
 
+func TestConfigDir(t *testing.T) {
+	colorGreen := "\033[32m"
+	colorReset := "\033[0m"
+
+	cwd, _ := os.Getwd()
+	fmt.Println(string(colorGreen), cwd, string(colorReset))
```
Comment (tom-hadlaw-hs) [inline config/config_test.go:81]: `t.Log/t.Logf` would be preferred here

```suggestion
	t.Log(string(colorGreen), cwd, string(colorReset))
```

File: config/config_test.go [hunk 10]
```diff
@@ -63,14 +65,31 @@ func TestInvalidConfigs(t *testing.T) {
...
 			}
 		})
 	}
 }
 
+func TestConfigDir(t *testing.T) {
+	colorGreen := "\033[32m"
+	colorReset := "\033[0m"
+
+	cwd, _ := os.Getwd()
```
Comment (tom-hadlaw-hs) [inline config/config_test.go:80]: errors not really relevant to the test, but imo I like to `panic(err)` any such errors. Might save someone some work debugging in the future.

Comment (tom-hadlaw-hs) [reply config/config_test.go:80]: or t.Fatal(err)

File: config/config_test.go [hunk 11]
```diff
@@ -63,14 +65,31 @@ func TestInvalidConfigs(t *testing.T) {
...
+	fmt.Println(string(colorGreen), cwd, string(colorReset))
+	for k, v := range validConfigs {
+		t.Run(k, func(t *testing.T) {
+			filename := mkConfig(t, v)
+			_, err := ReadConfigFile(filename, "./config.d", "", "")
+			if err != nil {
+				t.Fatalf("this config must be okay, got error: %v", err)
```
Comment (tom-hadlaw-hs) [inline config/config_test.go:87]: I think t.Fail would be a better fit here.

File: config/config.go [hunk 12]
```diff
@@ -168,7 +167,46 @@ func ReadConfigFile(configFile string, inputPrefix, outputPrefix string) (*Contr
...
+			currentConfigFile := util.AbsolutePath(inputPrefix, configDir) + "/" + item.Name()
+			log.Debug().Msg("reading config file %q", currentConfigFile)
+			yamlFile, err := ioutil.ReadFile(currentConfigFile)
+			if err != nil {
+				log.Error().Err(err).Msg("could not read config file")
+				return nil, fmt.Errorf("trouble reading config file %q: %w", currentConfigFile, err)
+			}
+
+			currentConfig, err := ReadConfig(log, yamlFile, inputPrefix, outputPrefix)
+			config.VaultConfig.Secrets = append(config.VaultConfig.Secrets, currentConfig.VaultConfig.Secrets...)
```
Comment (tom-hadlaw-hs) [inline config/config.go:195]: I think the behavior of this is a bit ambiguous. For example: this will add secret/templates/ssh-certs/aws-creds to the internal config but not top level config like the ConfigVersion.

I think this behavior makes sense, however we should explicitly document it (probably in the config dir flag help message)

File: config/config_test.go [hunk 13]
```diff
@@ -63,14 +65,31 @@ func TestInvalidConfigs(t *testing.T) {
...
+			_, err := ReadConfigFile(filename, "", "", "")
 			if err == nil {
 				t.Fatal("this config must generate an error")
 			}
 		})
 	}
 }
 
+func TestConfigDir(t *testing.T) {
+	colorGreen := "\033[32m"
```
Comment (tom-hadlaw-hs) [inline config/config_test.go:77]: Id prefer we keep the test output formatting standard. Also this will make jenkins build output a bit messy.

File: config/config_test.go [hunk 14]
```diff
@@ -63,14 +65,31 @@ func TestInvalidConfigs(t *testing.T) {
...
+	colorReset := "\033[0m"
+
+	cwd, _ := os.Getwd()
+	fmt.Println(string(colorGreen), cwd, string(colorReset))
+	for k, v := range validConfigs {
+		t.Run(k, func(t *testing.T) {
+			filename := mkConfig(t, v)
```
Comment (tom-hadlaw-hs) [inline config/config_test.go:84]: I feel like this test should specifically test the case where there are several config files in a config.d subdirectory.
mkConfig just creates a single file and returns the filename so this seems similar to the "ValidConfig" test above.

Perhaps we can modify mkConfig to take a list of body strings alongside the body, which if it is not empty will create additional files nested under "config.d" directory.

Then, after reading, check if data from all the files got correctly read and parsed.

Also would be worth adding an assertion on the behavior I was staging above (i.e. that ConfigVersion etc are invariant after the specified config rile) so that no one accidentally breaks that in the future.

Comment (tom-hadlaw-hs) [reply config/config_test.go:84]: You could also test the same code paths, but in the case where reading fails (probably less important since such a test already exists)

File: config/config.d/test-config.yml [hunk 15]
```diff
@@ -0,0 +1,8 @@
+version: 3
```
Comment (tom-hadlaw-hs) [inline config/config.d/test-config.yml:1]: what is this used for?

File: config/config_test.go (hunk 10 above)
Comment (marcial-lopezferrada-hs) [reply config/config_test.go:80]: sorry this was me troubleshooting my test :grimacing:

File: config/config_test.go (hunk 11 above)
Comment (marcial-lopezferrada-hs) [reply config/config_test.go:87]: t.Fail doesn't take any arguments though?

File: config/config.d/test-config.yml (hunk 15 above)
Comment (marcial-lopezferrada-hs) [reply config/config.d/test-config.yml:1]: that's the config file read from `config.d`

File: config/config.go (hunk 12 above)
Comment (marcial-lopezferrada-hs) [reply config/config.go:195]: gotcha, I'll add more info in the flag help

File: config/config.d/test-config.yml (hunk 15 above)
Comment (tom-hadlaw-hs) [reply config/config.d/test-config.yml:1]: is it used for testing?

File: config/config.go (hunk 3 above)
Comment (tom-hadlaw-hs) [reply config/config.go:155]: yeah makes sense. Especially when you consider the top level configuration contained in it.

File: config/config_test.go (hunk 11 above)
Comment (tom-hadlaw-hs) [reply config/config_test.go]: usually we do t.Log followed by t.Fail.

File: config/config_test.go [hunk 16]
```diff
@@ -63,14 +63,43 @@ func TestInvalidConfigs(t *testing.T) {
...
 }
 
+func TestConfigDir(t *testing.T) {
+
+	for k, v := range validConfigs {
+		t.Run(k, func(t *testing.T) {
+			filename := mkConfig(t, v)
+			config, err := ReadConfigFile(filename, "./config.d", "", "")
+			if err != nil {
+				// t.Fail("this config must be okay, got error")
```
Comment (tom-hadlaw-hs) [inline config/config_test.go:81]: should remove commented code

File: config/config.d/test-config.yml (hunk 15 above)
Comment (tom-hadlaw-hs) [reply config/config.d/test-config.yml:1]: oh nm I see. Can we have these as strings in code that we write to the temp directory here: https://github.com/hootsuite/vault-ctrl-tool/blob/ffabbe0ce621e0e5fefc79f93ff4f8bc0f67758d/config/config_test.go#L104

File: config/config_test.go [hunk 17]
```diff
@@ -62,17 +78,47 @@ func TestValidConfigs(t *testing.T) {
...
 		})
 	}
 }
 
-func mkConfig(t *testing.T, body string) string {
-	f, err := ioutil.TempFile(t.TempDir(), "config_test_*")
+func TestConfigDir(t *testing.T) {
+
+	dir := t.TempDir()
+	f := mkConfig(t, dir, validSubConfig["sub config1"])
```
Comment (tom-hadlaw-hs) [inline config/config_test.go:93]: You could probably make this a bit cleaner and generalized by just iterating over the map:

```suggestion
	for configName, configData := range validSubConfig {
		f := mkConfig(t, dir, configData)
		t.Logf("created config for: %s: %s", configName, f)
	} 
```

Comment (tom-hadlaw-hs) [reply config/config_test.go:93]: + removing calls to mkConfig/t.Log bellow

File: config/config_test.go [hunk 18]
```diff
@@ -62,17 +78,47 @@ func TestValidConfigs(t *testing.T) {
...
+	if err != nil {
+		t.Log("this config must be okay, got error")
+		t.Fail()
+	}
+	// check if key == test exists
+	// check if key == test2 exists
+	foundKey1 := false
+	foundKey2 := false
+	for _, secret := range config.VaultConfig.Secrets {
+		if secret.Key == "test" {
```
Comment (tom-hadlaw-hs) [inline config/config_test.go:108]: For completeness, can we check the other keys

File: config/config_test.go [hunk 19]
```diff
@@ -62,17 +78,47 @@ func TestValidConfigs(t *testing.T) {
...
+	config, err := ReadConfigFile(emptyConfig, dir, "", "")
+	if err != nil {
+		t.Log("this config must be okay, got error")
+		t.Fail()
+	}
+	// check if key == test exists
+	// check if key == test2 exists
+	foundKey1 := false
+	foundKey2 := false
+	for _, secret := range config.VaultConfig.Secrets {
```
Comment (tom-hadlaw-hs) [inline config/config_test.go:107]: Would it be worth doing with a different config type as well to make sure it works when there's say Secrets + AWS in the same file or spread across multiple files.

File: config/config_test.go [hunk 20]
```diff
@@ -62,17 +78,47 @@ func TestValidConfigs(t *testing.T) {
...
+		if secret.Key == "test" {
+			foundKey1 = true
+		}
+		if secret.Key == "test2" {
+			foundKey2 = true
+		}
+	}
+	if !foundKey1 || !foundKey2 {
+		t.Fatal("failed to find keys in config directory")
```
Comment (tom-hadlaw-hs) [inline config/config_test.go:116]: I think we should use Fail here

File: config/config_test.go [hunk 21]
```diff
@@ -62,17 +78,47 @@ func TestValidConfigs(t *testing.T) {
...
 			}
 		})
 	}
 }
 
-func mkConfig(t *testing.T, body string) string {
-	f, err := ioutil.TempFile(t.TempDir(), "config_test_*")
+func TestConfigDir(t *testing.T) {
+
+	dir := t.TempDir()
```
Comment (tom-hadlaw-hs) [inline config/config_test.go:92]: This is more just about how I might approach this so don't feel like this needs to be done.

But for testing functionality like this where you're testing the same basic functionality but trying to capture different scenarios and edge cases it would be good to use table driven tests where you can easily add more scenarios.

File: config/config_test.go (hunk 20 above)
Comment (tom-hadlaw-hs) [reply config/config_test.go:116]: Every other package test in this repo uses: https://github.com/stretchr/testify for its assertions. I wonder why this package only uses the standard testing library.

File: config/config_test.go (hunk 21 above)
Comment (tom-hadlaw-hs) [reply config/config_test.go:92]: I would do something like define a struct that describes a single test:

```go
type configDirTest struct {
    files []string
    expectedSecrets []config.SecretType
    expectedAWS []config.AWSType
    expectedErr error
}
```

Then you can define your tests in a table:
```
var configDirTests = map[string]configDirTest{
    "test happy path!": {
        files: []string{
version: 3
secrets:
 - key: test
   path: /secret/test
   lifetime: static`,
version: 3
secrets:
 - key: test2
   path: /secret/test2
   lifetime: static
`
        },
        expectedSecrets: []config.SecretType{{Key: "test2"}, {Key: "test"}},
        expectedErr: nil,
    },
    // ...
}
```

And then finally you could use the assert library this project uses for other tests to check if the final config contains your expected objects: https://pkg.go.dev/github.com/stretchr/testify@v1.7.0/assert#Assertions.Contains

(See my comment about using the assert function regardless)

File: config/config_test.go (hunk 18 above)
Comment (tom-hadlaw-hs) [reply config/config_test.go:108]: Maybe we could use https://pkg.go.dev/github.com/stretchr/testify@v1.7.0/assert#Assertions.Contains to check if the output array contains your expected values.

File: config/config_test.go (hunk 21 above)
Comment (tom-hadlaw-hs) [reply config/config_test.go:92]: Then you could also do stuff like test cases where the config reading fails using the expected error.

File: config/config_test.go [hunk 22]
```diff
@@ -62,17 +78,47 @@ func TestValidConfigs(t *testing.T) {
...
+			foundKey2 = true
+		}
+	}
+	if !foundKey1 || !foundKey2 {
+		t.Fatal("failed to find keys in config directory")
+	}
+}
+
+func mkConfig(t *testing.T, directory string, body string) string {
+	f, err := ioutil.TempFile(directory, "config_test_*.yml")
```
Comment (tom-hadlaw-hs) [inline config/config_test.go:121]: why are we naming the file with _*?

File: util/flags.go [hunk 23]
```diff
@@ -103,8 +104,9 @@ func ProcessFlags(args []string) (*CliFlags, error) {
 
 	app.Flag("init", "Run in init mode, process templates and exit.").Default("false").BoolVar(&flags.PerformInit)
 	app.Flag("config", "Full path of the config file to read.").Default("vault-config.yml").StringVar(&flags.ConfigFile)
+	app.Flag("config-dir", "Full path of the config directory to read config files. Be aware that the config version is read only once so all sub config should have the same version.").Default("config.d").StringVar(&flags.ConfigDir)
 	app.Flag("output-prefix", "Path to prefix to all output files (such as /etc/secrets)").StringVar(&flags.OutputPrefix)
-	app.Flag("input-prefix", "Path to prefix on all files being read; including the config file.").StringVar(&flags.InputPrefix)
+	app.Flag("input-prefix", "Path to prefix on all files being read; including the config file and config directory.").StringVar(&flags.InputPrefix)
```
Comment (tom-hadlaw-hs) [inline util/flags.go:109]: Can we also specify that files found in the input prefix that aren't specified using the config-file flag won't have their root values used?

File: config/config_test.go (hunk 22 above)
Comment (marcial-lopezferrada-hs) [reply config/config_test.go:121]: it was like that when I showed up, but I can change it to a camelcase style

Comment (tom-hadlaw-hs) [reply config/config_test.go:121]: oh no thats fine, I was just wondering.

File: config/config_test.go (hunk 21 above)
Comment (marcial-lopezferrada-hs) [reply config/config_test.go:92]: oh I see, yeah that sounds great. I'll come back to it though to make some progress

File: config/config.go [hunk 24]
```diff
@@ -168,7 +167,46 @@ func ReadConfigFile(configFile string, inputPrefix, outputPrefix string) (*Contr
...
 
-	return ReadConfig(log, yamlFile, inputPrefix, outputPrefix)
+	config, err := ReadConfig(log, yamlFile, inputPrefix, outputPrefix)
+
+	if err != nil {
+		log.Error().Err(err).Msg("could not parse config file")
+		return nil, fmt.Errorf("could not parse config file %q: %w", absConfigFile, err)
+	}
+
+	items, err := ioutil.ReadDir(util.AbsolutePath(inputPrefix, configDir))
```
Comment (tom-hadlaw-hs) [inline config/config.go:177]: If configDir is empty, I don't think we should attempt to read in config files

Comment (tom-hadlaw-hs) [reply config/config.go:177]: (just default to old behaviour then)
